resources/js/mp-history/main.tsx

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import HeaderV4 from 'components/header-v4';
+import LegacyMatchEventJson from 'interfaces/legacy-match-event-json';
+import LegacyMatchJson from 'interfaces/legacy-match-json';
+import UserJson from 'interfaces/user-json';
+import { route } from 'laroute';
+import { keyBy, dropRightWhile } from 'lodash';
+import * as React from 'react';
+import { classWithModifiers } from 'utils/css';
+import Content from './content';
+
+interface LegacyMultiplayerHistoryJson {
+  current_game_id?: number;
+  events: LegacyMatchEventJson[];
+  first_event_id: number;
+  latest_event_id: number;
+  match: LegacyMatchJson;
+  users: UserJson[];
+}
+
+interface Props {
+  events: LegacyMultiplayerHistoryJson;
+}
+
+interface State {
+  currentGameId?: number;
+  events: LegacyMatchEventJson[];
+  latestEventId: number;
+  loadingNext: boolean;
+  loadingPrevious: boolean;
+  match: LegacyMatchJson;
+  users: Partial<Record<number, UserJson>>;
+}
+
+const FETCH_LIMIT = 100;
+const MAXIMUM_EVENTS = 500;
+const REFRESH_TIMEOUT = 10000;
+
+export default class Main extends React.Component<Props, State> {
+  private readonly loadNext: () => void;
+  private readonly loadPrevious: () => void;
+
+  private readonly timeouts: Partial<Record<string, number>> = {};
+
+  constructor(props: Props) {
+    super(props);
+
+    const events = props.events.events;
+
+    this.state = {
+      currentGameId: props.events.current_game_id,
+      events,
+      latestEventId: props.events.latest_event_id,
+      loadingNext: false,
+      loadingPrevious: false,
+      match: props.events.match,
+      users: keyBy(props.events.users, 'id'),
+    };
+
+    this.loadNext = () => {
```

Comment:
you can do this at class definition level...
```typescript
class What {
  private readonly o_O = () => {
    ...
  };
}
```

Reply:
touché